nobrainer/processing/generation.py

```diff
@@ -0,0 +1,191 @@
+        mod = importlib.import_module("..models", "nobrainer.processing")
+        base_model = getattr(mod, self.base_model)
+        _create(base_model)
+        print(self.generator_.summary())
+        print(self.discriminator_.summary())
+
+        for resolution in self.resolutions_:
+
+            # create a train dataset with features for resolution
+            dataset_train = get_dataset(
+                file_pattern=self.file_pattern % (resolution),
+                batch_size=self.resolution_batch_size_map_[resolution],
+                num_parallel_calls=num_parallel_calls,
+                volume_shape=(resolution, resolution, resolution),
+                n_classes=n_classes,
+                scalar_label=True,
+                normalizer=None,
+            )
+
+            # grow the networks by one (2^x) resolution
+            self.generator_.add_resolution()
+            self.discriminator_.add_resolution()
+
+            # Note: multigpu should be added here
+            # compile the trainers
+            _compile()
+
+            steps_per_epoch = epochs // self.resolution_batch_size_map_[resolution]
+            # save_best_only is set to False as it is an adversarial loss
+            model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
+                str(model_dir),
+                save_weights_only=True,
+                save_best_only=False,
+                save_freq=10,
+            )
+
+            # Train at resolution
+            print("Resolution : {}".format(resolution))
+
+            print("Transition phase")
+            self.pgan_trainer.fit(
+                dataset_train,
+                phase="transition",
+                resolution=resolution,
+                steps_per_epoch=steps_per_epoch,  # necessary for repeat dataset
+                callbacks=[model_checkpoint_callback],
+            )
+
+            print("Resolution phase")
+            self.pgan_trainer.fit(
+                dataset_train,
+                phase="resolution",
+                resolution=resolution,
+                steps_per_epoch=steps_per_epoch,
+                callbacks=[model_checkpoint_callback],
+            )
+
+            # save the final weights
+            self.generator_.save(
+                str(model_dir.joinpath("generator_res_{}".format(resolution)))
+            )
+
+        return self
+
+    def generate(
+        self,
+        x,
+        latent_size=1024,
+    ):
+        """generate a synthetic image using the trained model"""
+        if self.model_ is None:
+            raise ValueError("Model is undefined. Please train or load a model")
+        import nibabel as nib
+        import numpy as np
+
+        latents = tf.random.normal((1, latent_size))
+        generate = self.generator_.signatures["serving_default"]
+        img = generate(latents)["generated"]
+        img = np.squeeze(img)
+        img = nib.Nifti1Image(img.astype(np.uint8), np.eye(4))
```

Comment:
for this kind of model i think uint16 is better.

Reply:
I will fix it.